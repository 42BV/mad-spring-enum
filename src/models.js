// @flow

export type EnumValues = Array<string>;

/**
 * The enums should have the following signature:
 *
 * {
 *   "UserRole": [
 *     "ADMIN",
 *     "USER"
 *  ],
 *  "BillingInterval": [
 *    "MONTHLY", "WEEKLY", "DAILY"
 *  ]
 * }
 *
 * The keys represent the name of the enum, in the above case 'UserRole',
 * and 'BillingInterval'. The array points to the possible values 
 * of the enum.
 */
export type Enums = {
  [key: string]: EnumValues
};

