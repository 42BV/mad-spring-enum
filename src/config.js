// @flow

import type { EnumsStore, Action } from './enums-reducer';

export type Config = {
  // The URL which will provide the enums over a GET request.
  enumsUrl: string,

  // Whether or not the 'enumsUrl' should be called with authentication.
  needsAuthentication: boolean,

  // The dispatch function for the Redux store.
  dispatch: (action: Action) => void,

  // A function which returns the latests EnumsStore from Redux.
  enumsStore: () => EnumsStore
};

let config: Config | null = null;

/**
 * Configures the Constraint libary.
 *
 * @param {Config} The new configuration
 */
export function configureEnums(c: Config) {
  config = c;
}

/**
 * Either returns the a Config or throws an error when the
 * config is not yet initialized.
 *
 * @returns The Config
 */
export function getConfig(): Config {
  if (config === null) {
    throw new Error('The enum service is not initialized.');
  } else {
    return config;
  }
}
