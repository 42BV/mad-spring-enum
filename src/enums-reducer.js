// @flow

import type { Enums } from './models';

export type Action = 
  | { type: 'MAD_SPRING_ENUM.SET_ENUMS', enums: Enums };

export const SET_ENUMS = 'SET_ENUMS';

export type EnumsStore = {
  +enums?: Enums,
};

export const initialState: EnumsStore = {
  enums: undefined,
};

export function enums(state: EnumsStore = initialState, action: Action): EnumsStore {
  switch(action.type) {
    case 'MAD_SPRING_ENUM.SET_ENUMS': {
      return { ...state, enums: action.enums };
    }

    default: {
      return state;
    }
  }
}

export function setEnums(enums: Enums): Action {
  return { type: 'MAD_SPRING_ENUM.SET_ENUMS', enums };
}
