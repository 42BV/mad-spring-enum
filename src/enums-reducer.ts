import { Action, Reducer } from 'redux';
import { Enums } from './models';

export const SET_ENUM = 'MAD_SPRING_ENUM.SET_ENUMS';

export interface SetEnumAction extends Action<typeof SET_ENUM> {
  enums: Enums;
}

export type EnumActions = SetEnumAction;

export interface EnumsStore {
  enums: Enums;
}

export const initialState: EnumsStore = {
  enums: {},
};

export const enums: Reducer<EnumsStore, EnumActions> = (
  state: EnumsStore = initialState,
  action: EnumActions,
): EnumsStore => {
  switch (action.type) {
    case SET_ENUM: {
      return { ...state, enums: action.enums };
    }

    default: {
      return state;
    }
  }
};

export function setEnums(enums: Enums): SetEnumAction {
  return { type: 'MAD_SPRING_ENUM.SET_ENUMS', enums };
}
