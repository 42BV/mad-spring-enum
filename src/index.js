// @flow

export { configureEnums } from './config.js';
export { enums } from './enums-reducer.js';
export { loadEnums } from './load-enums.js';
export { getEnum } from './selectors';

export type { Enums, EnumValues } from './models.js';
export type { EnumsStore } from './enums-reducer.js';
