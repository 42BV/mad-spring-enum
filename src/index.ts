export { configureEnums, Config } from './config';
export { enums } from './enums-reducer';
export { loadEnums } from './load-enums';
export { getEnum } from './selectors';
export { EnumsProvider, useEnums, EnumsContext } from './provider';
export { Enums, EnumValues } from './models';
