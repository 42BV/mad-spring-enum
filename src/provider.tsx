import React, { createContext, useContext, useState } from 'react';
import { Enums } from './models';

interface Props {
  enums: Enums;
}
export const EnumsContext = createContext<Enums>({});

export function useEnums(): Enums {
  return useContext(EnumsContext);
}

export const EnumsProvider: React.FC<Props> = ({ enums, children }) => {
  const [state] = useState(enums);
  return <EnumsContext.Provider value={{ ...state }}>{children}</EnumsContext.Provider>;
};
