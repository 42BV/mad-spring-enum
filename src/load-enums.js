
import { setEnums } from './enums-reducer';

import { getConfig } from './config';

/**
 * Loads the enums from the back-end.
 *
 * The URL it will send the request to is defined by the 'enumsUrl'
 * from the Config object. The HTTP method it uses is 'get'.
 *
 * It will also send the credentials if 'needsAuthentication' from
 * the Config object is set to true.
 *
 * The entire response will be written to the Redux EnumsStore's
 * enums key. Whatever the JSON response is will be the enums.
 *
 *  An example response:
 *
 * ```JSON
 * {
 *   "UserRole": [
 *     "ADMIN",
 *     "USER"
 *  ],
 *  "BillingInterval": [
 *    "MONTHLY", "WEEKLY", "DAILY"
 *  ]
 * ```
 *
 * @returns {Promise}
 */
export function loadEnums(): Promise<*> {
  const { enumsUrl, needsAuthentication, dispatch } = getConfig();

  const config = needsAuthentication ? { credentials: 'include'} : {};

  return fetch(enumsUrl, config)
    .then(tryParse)
    .then((enums) => {
      dispatch(setEnums(enums));
    });
}

// Throw error when not 200 otherwise parse response.
function tryParse(response: Response) {
  if (response.status !== 200) {
    throw response;
  } else {
    return response.json();
  }
}
