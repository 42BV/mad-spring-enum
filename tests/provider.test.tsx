import React from 'react';
import { EnumsProvider, useEnums, EnumsContext } from '../src/provider';
import renderer from 'react-test-renderer';

const HookTest = () => {
  const { CAR_TYPES } = useEnums();
  return <h2>I really like {CAR_TYPES[2]} cars</h2>;
};

class ConsumerTest extends React.Component<{ favoriteCar: string }> {
  render() {
    const { favoriteCar } = this.props;
    return <h2>My favorite car brand is {favoriteCar}</h2>;
  }
}

describe('EnumProvider', () => {
  test('should provide context as hook', () => {
    const tree = renderer
      .create(
        <EnumsProvider enums={{ CAR_TYPES: ['AUDI', 'MUSTANG', 'TESLA'] }}>
          <HookTest />
        </EnumsProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('should provide context as consumer', () => {
    const tree = renderer
      .create(
        <EnumsProvider enums={{ CAR_TYPES: ['AUDI', 'MUSTANG', 'TESLA'] }}>
          <EnumsContext.Consumer>
            {({ CAR_TYPES }) => <ConsumerTest favoriteCar={CAR_TYPES[1]} />}
          </EnumsContext.Consumer>
        </EnumsProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
