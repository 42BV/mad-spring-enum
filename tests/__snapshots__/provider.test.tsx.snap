// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnumProvider should provide context as consumer 1`] = `
<h2>
  My favorite car brand is 
  MUSTANG
</h2>
`;

exports[`EnumProvider should provide context as hook 1`] = `
<h2>
  I really like 
  TESLA
   cars
</h2>
`;
